syllabus/core/stat_recorder.py

```diff
@@ -38,37 +42,25 @@ def record(self, episode_return: float, episode_length: int, episode_task, env_i
 
         if episode_task in self.tasks:
             if self.calc_past_N:
-                self.records[episode_task].append({
-                    "r": episode_return,
-                    "l": episode_length,
-                    "env_id": env_id
-                })
-                self.records[episode_task] = self.records[episode_task][-self.calc_past_N:]
-                self.stats[episode_task]['mean_r'] = np.mean([record["r"] for record in self.records[episode_task]])
-                self.stats[episode_task]['var_r'] = np.var([record["r"] for record in self.records[episode_task]])
-                self.stats[episode_task]['mean_l'] = np.mean([record["l"] for record in self.records[episode_task]])
-                self.stats[episode_task]['var_l'] = np.var([record["l"] for record in self.records[episode_task]])
-                # only save mean/variance the past N episodes
+                self.episode_returns[episode_task].append(episode_return)
+                self.episode_lengths[episode_task].append(episode_length)
+                self.env_ids[episode_task].append(env_id)
+
+                self.stats[episode_task]['mean_r'] = np.mean(list(self.episode_returns[episode_task])[-self.calc_past_N:]) # I am not sure whether there is a more efficient way to slice to deque. I temperorily convert it to a list then slice it, which should cost O(n)
```

Comment:
You shouldn't need to slice a deque at all, it should automatically drop elements when it goes past keep_last_n. Check the documentation for it